File: dekaf-core/src/main/java/org/jetbrains/dekaf/core/BaseScriptRunner.java
package org.jetbrains.dekaf.core;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.dekaf.sql.SqlCommand;
import org.jetbrains.dekaf.sql.SqlQuery;
import org.jetbrains.dekaf.sql.SqlScript;
import org.jetbrains.dekaf.sql.SqlStatement;

import java.util.List;



/**
 * @author Leonid Bushuev from JetBrains
 **/
public class BaseScriptRunner implements DBScriptRunner {

  @NotNull
  private final DBTransaction myTransaction;

  @NotNull
  private final SqlScript myScript;


  BaseScriptRunner(@NotNull final DBTransaction transaction,
                   @NotNull final SqlScript script) {
    myTransaction = transaction;
    myScript = script;
  }


  @Override
  public DBScriptRunner run() {
    List<? extends SqlStatement> statements = myScript.getStatements();

    for (SqlStatement statement : statements) {
      if (statement instanceof SqlCommand) {
        runCommand((SqlCommand)statement);
      }
      if (statement instanceof SqlQuery) {
        runQuery((SqlQuery) statement);
      }
    }

    return this;
  }

  private void runCommand(final SqlCommand command) {
    myTransaction.command(command).run();
  }

  private void runQuery(final SqlQuery query) {
    myTransaction.query(query).run();
  }


}

File: dekaf-main/src/impl/BaseScriptRunner.kt
package org.jetbrains.dekaf.impl

import org.jetbrains.dekaf.core.DBScriptRunner
import org.jetbrains.dekaf.sql.SqlCommand
import org.jetbrains.dekaf.sql.SqlQuery
import org.jetbrains.dekaf.sql.SqlScript

/**
 * Script runner.
 */
internal class BaseScriptRunner : BaseRunner, DBScriptRunner {

    private val script: SqlScript


    constructor(session: BaseSession, script: SqlScript) : super(session) {
        this.script = script
    }


    override fun run(): BaseScriptRunner {
        val n = script.count()

        // make runners
        val runners = Array<BaseStatementRunner>(n) { index ->
            val statement = script.statements[index]
            when (statement) {
                is SqlQuery<*> -> session.query(statement)
                is SqlCommand  -> session.command(statement)
                else           -> throw IllegalArgumentException("Unknown how to deal with class "+statement.javaClass.name)
            }
        }

        // run
        for (runner in runners) {
            try {
                runner.prepare()
                runner.run()
            }
            finally {
                runner.close()
            }
        }

        // ok
        return this
    }


    override fun close() {}

}